tidy3d/plugins/near2far/near2far.py

```diff
@@ -1,115 +1,373 @@
+        sim_data: SimulationData,
+        mons: List['FieldMonitor'],
+        normal_dirs: List[Direction],
+        frequency: float,
+        pts_per_wavelength: int = 10,
+        medium: Medium = None):
+        """Constructs :class:`Near2Far` from a list of surface monitors and their directions
 
-        # grid sizes
-        self.dx = sim_data.simulation.grid_size[0]
-        self.dy = sim_data.simulation.grid_size[1]
+        Parameters
+        ----------
+        sim_data : :class:`.SimulationData`
+            Container for simulation data containing the near field monitors.
+        mons : List[:class:`.FieldMonitor`]
+            List of :class:`.FieldMonitor` objects on which near fields will be sampled.
+        normal_dirs : List[:class:`.Direction`]
+            List containing the :class:`.Direction` of the normal to each surface monitor 
+            w.r.t. to the positive x, y or z unit vectors. Must have the same length as mons.
+        frequency : float
+            Frequency to select from each :class:`.FieldMonitor` to use for projection.
+            Must be a frequency stored in each :class:`FieldMonitor`.
+        pts_per_wavelength : int
+            Number of points per wavelength with which to discretize the
+            surface monitors for the projection. Default: 10.
+        medium : :class:`.Medium`
+            Background medium in which to radiate near fields to far fields.
+            Default: same as the :class:`.Simulation` background medium.
+        """
 
-        # get coordinate at centers
-        x_centers = field_data.x.values
-        y_centers = field_data.y.values
-        self.xx, self.yy = np.meshgrid(x_centers, y_centers, indexing="ij")
+        if len(mons) != len(normal_dirs):
+            raise SetupError("Number of monitors does not equal the number of directions.")
 
-        # compute equivalent sources
-        self.Jx = -np.squeeze(Hy.values)
-        self.Jy = np.squeeze(Hx.values)
-        self.Mx = np.squeeze(Ey.values)
-        self.My = -np.squeeze(Ex.values)
+        surfaces = []
+        for mon, normal_dir in zip(mons, normal_dirs):
+            surfaces.append(Near2FarSurface(sim_data=sim_data, mon=mon, normal_dir=normal_dir))
 
-    def _radiation_vectors(self, theta: float, phi: float):
-        """Compute radiation vectors at an angle in spherical coordinates
+        return cls(
+            sim_data=sim_data,
+            surfaces=surfaces,
+            frequency=frequency,
+            pts_per_wavelength=pts_per_wavelength,
+            medium=medium
+            )
+
+    def initialize(self):
+        """Precomputes surface currents and other data in preparation for a near-to-far projection.
+            Will be called automatically the first time any far field computation is performed.
+        """
+
+        if self.medium is None:
+            self.medium = self.sim_data.simulation.medium
+
+        wavelength = C_0 / self.frequency / self.medium.permittivity
```

Comment:
I dont think this is correct.  Shouldn't it be index? not permittivity?

Reply:
Thanks for catching that, yeah there's a missing square root

Reply:
rather than square root, let's use that `eps_complex_to_nk` method just to keep all of that in the same place

Reply:
in fact, in this case you could do
```python
eps_complex = self.medium.eps_model(self.frequency)
n, k = self.medium.eps_complex_to_nk(eps_complex)
wavelength = C_0 / self.frequency / n
```
or something like this